### ISSUE TITLE
OC\Files\Stream\Encryption::loadContext($name) should be compatible with Icewind\Streams\Wrapper::loadContext

### ISSUE THREAD

**Issue by andyxheli** [nextcloud/server#26464]
@vitormattos 

Getting the followung error once updated to NC 20.0.9 everytime i open a file or scroll throw the photos app or files 

#### Nextcloud Error 1 (data/nextcloud.log)
<details>
<summary>Error: Declaration of OC\Files\Stream\Encryption::wrapSource($source, $context, $protocol, $class, $mode = 'r+') should be compatible with Icewind\Streams\WrapperHandler::wrapSource($source, $context = Array, $protocol = NULL, $class = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#201</summary>

```
{"reqId":"UTnfI0uqU071I3r5FXI7","level":3,"time":"2021-04-08T16:21:26+00:00","remoteAddr":"","user":"","app":"PHP","method":"GET","url":"/apps/onlyoffice/download?doc=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJhY3Rpb24iOiJkb3dubG9hZCIsImZpbGVJZCI6Mzk4OTkzLCJ1c2VySWQiOiJheGhlbGkifQ.A1eXyrt0TLUYfhhEhgdru1bEGfnBESxs55v62iMM--0","message":{"Exception":"Error","Message":"Declaration of OC\\Files\\Stream\\Encryption::wrapSource($source, $context, $protocol, $class, $mode = 'r+') should be compatible with Icewind\\Streams\\WrapperHandler::wrapSource($source, $context = Array, $protocol = NULL, $class = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#201","Code":0,"Trace":[{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":38,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/lib/composer/composer/ClassLoader.php","line":444,"args":["/var/www/nextcloud/lib/private/Files/Stream/Encryption.php"],"function":"include"},{"file":"/var/www/nextcloud/lib/composer/composer/ClassLoader.php","line":322,"function":"Composer\\Autoload\\includeFile"},{"function":"loadClass","class":"Composer\\Autoload\\ClassLoader","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":469,"function":"spl_autoload_call"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":218,"function":"fopen","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":246,"function":"file_get_contents","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":1167,"function":"file_get_contents","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":597,"function":"basicOperation","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Node/File.php","line":57,"function":"file_get_contents","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/apps/onlyoffice/controller/callbackcontroller.php","line":295,"function":"getContent","class":"OC\\Files\\Node\\File","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"download","class":"OCA\\Onlyoffice\\Controller\\CallbackController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":309,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1008,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Node.js/6.13","version":"20.0.9.1","id":"606f2e1fb6e10"}
```
</details>

#### Nextcloud Error 2 (data/nextcloud.log)
<details>
<summary>Error: Declaration of OC\Files\Stream\Encryption::loadContext($name) should be compatible with Icewind\Streams\Wrapper::loadContext($name = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#224
--

</summary>

```
{"reqId":"UTnfI0uqU071I3r5FXI7","level":3,"time":"2021-04-08T16:21:26+00:00","remoteAddr":"","user":"","app":"PHP","method":"GET","url":"/apps/onlyoffice/download?doc=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJhY3Rpb24iOiJkb3dubG9hZCIsImZpbGVJZCI6Mzk4OTkzLCJ1c2VySWQiOiJheGhlbGkifQ.A1eXyrt0TLUYfhhEhgdru1bEGfnBESxs55v62iMM--0","message":{"Exception":"Error","Message":"Declaration of OC\\Files\\Stream\\Encryption::loadContext($name) should be compatible with Icewind\\Streams\\Wrapper::loadContext($name = NULL) at /var/www/nextcloud/lib/private/Files/Stream/Encryption.php#224","Code":0,"Trace":[{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":38,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::"},{"file":"/var/www/nextcloud/lib/composer/composer/ClassLoader.php","line":444,"args":["/var/www/nextcloud/lib/private/Files/Stream/Encryption.php"],"function":"include"},{"file":"/var/www/nextcloud/lib/composer/composer/ClassLoader.php","line":322,"function":"Composer\\Autoload\\includeFile"},{"function":"loadClass","class":"Composer\\Autoload\\ClassLoader","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":469,"function":"spl_autoload_call"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":218,"function":"fopen","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":246,"function":"file_get_contents","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":1167,"function":"file_get_contents","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":597,"function":"basicOperation","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Node/File.php","line":57,"function":"file_get_contents","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/apps/onlyoffice/controller/callbackcontroller.php","line":295,"function":"getContent","class":"OC\\Files\\Node\\File","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":169,"function":"download","class":"OCA\\Onlyoffice\\Controller\\CallbackController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":100,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":152,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":309,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":1008,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":37,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/lib/private/Log/ErrorHandler.php","Line":91,"CustomMessage":"--"},"userAgent":"Node.js/6.13","version":"20.0.9.1","id":"606f2e1fb6eb0"}
```
</details>

**Comment by axheli** [comment]
@jknockaert  I compared the files between NC 20.0.9 and 21.0.1 and i see a code change on NC 21.0.1 and on NC 20.0.9 thats missing would this have anything to do with it ? 

\lib\private\Files\Stream\Encryption.php
![image](https://user-images.githubusercontent.com/59488153/114091139-e9749e00-987d-11eb-92f8-5be74ab69d4f.png)

![image](https://user-images.githubusercontent.com/59488153/114091243-09a45d00-987e-11eb-8550-2412bdfe0c0b.png)

**Comment by axheli** [comment]
This also seems to happen on Nextcloud 21.0.1

**Comment by Zaxooy** [comment]
@jknockaert You asked in #24966 to get tagged for this. 🙂

**Comment by jknockaert** [comment]
@axheli Error 3 is a duplicate of #26473

**Comment by andyxheli** [comment]
@jknockaert did you have anytime to look into this :)

**Comment by jknockaert** [comment]
Yes. Error 3 is a duplicate of #26473. For the other two errors I can't yet make sense of it. I'd rather first have a better understanding of what we see happening in some other recent errors which may be wrapper related.

**Comment by andyxheli** [comment]
Thank you! @jknockaert is there anything you need on my end ? more logs ?

**Comment by jknockaert** [comment]
Are you actually using SMB storage for your files?

**Comment by andyxheli** [comment]
Nope everything is local i dont use SMB storage

**Comment by jknockaert** [comment]
OK let's keep this one open for now.

**Comment by andyxheli** [comment]
Okay sounds good!

**Comment by jknockaert** [comment]
@andyxheli Today I decided to switch off encryption on my NC server, so I will no longer investigate any issues relating to its sloppy implementation.

**Comment by andyxheli** [comment]
@jknockaert Thanks for your time i think im going to do the same nothing but issues and i don't know what everything work perfect on Nextcloud 20 and everything is all messed up on Nextcloud 21 crazy.

**Comment by andyxheli** [comment]
@vitormattos anyway you could help with this please? :)

**Comment by appzer0** [comment]
This issue is referenced as solution on NC's forum, though:

I upgraded to NC 21.0.2 RC1 and the upload fails are still there: 

```
[PHP] Error: Error: Undefined index: encryptedVersion at /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php#669 at <<closure>>

 0. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php line 669
    OC\Log\ErrorHandler::onError(8, "Undefined index: encryptedVersion", "/var/www/nextcl ... p", 669, {sourceStorage:  ... }})
 1. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php line 767
    OC\Files\Storage\Wrapper\Encryption->updateEncryptedVersion("*** sensitive parameters replaced ***")
 2. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php line 623
    OC\Files\Storage\Wrapper\Encryption->copyBetweenStorage("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 3. /var/www/nextcloud/apps/files_trashbin/lib/Storage.php line 240
    OC\Files\Storage\Wrapper\Encryption->moveFromStorage("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 4. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/File.php line 292
    OCA\Files_Trashbin\Storage->moveFromStorage("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 5. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php line 156
    OCA\DAV\Connector\Sabre\File->put(null)
 6. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 1098
    OCA\DAV\Connector\Sabre\Directory->createFile("logo stop covid 2.jpg", null)
 7. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 504
    Sabre\DAV\Server->createFile("files/Romain J/ ... g", null, null)
 8. /var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php line 89
    Sabre\DAV\CorePlugin->httpPut(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})
 9. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 472
    Sabre\DAV\Server->emit("method:PUT", [Sabre\HTTP\Requ ... }])
10. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 253
    Sabre\DAV\Server->invokeMethod(Sabre\HTTP\Request {}, Sabre\HTTP\Response {})
11. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 321
    Sabre\DAV\Server->start()
12. /var/www/nextcloud/apps/dav/lib/Server.php line 332
    Sabre\DAV\Server->exec()
13. /var/www/nextcloud/apps/dav/appinfo/v2/remote.php line 35
    OCA\DAV\Server->exec()
14. /var/www/nextcloud/remote.php line 167
    require_once("/var/www/nextcl ... p")

PUT /remote.php/dav/files/*sensitive*.jpg
from XXXXXXXX by XXXXX at 2021-05-17T11:31:52+02:00
```

**Comment by andyxheli** [comment]
Still an issue on Nextcloud 20.0.10

**Comment by andyxheli** [comment]
@rullzer im really sorry to bother you I was just wondering do you have any feedback on this?

**Comment by Obi-Wahn** [comment]
Same error with 21.0.2
`Error: Undefined index: encryptedVersion at /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php#669
/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php - line 669:

OC\Log\ErrorHandler::onError()

/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php - line 767:

OC\Files\Storage\Wrapper\Encryption->updateEncryptedVersion("*** sensiti ... *")

/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php - line 623:

OC\Files\Storage\Wrapper\Encryption->copyBetweenStorage()

/var/www/nextcloud/apps/files_trashbin/lib/Storage.php - line 240:

OC\Files\Storage\Wrapper\Encryption->moveFromStorage()

/var/www/nextcloud/apps/dav/lib/Connector/Sabre/File.php - line 292:

OCA\Files_Trashbin\Storage->moveFromStorage()

/var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php - line 156:

OCA\DAV\Connector\Sabre\File->put()

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 1098:

OCA\DAV\Connector\Sabre\Directory->createFile()

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php - line 504:

Sabre\DAV\Server->createFile()

/var/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php - line 89:

Sabre\DAV\CorePlugin->httpPut()

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 472:

Sabre\DAV\Server->emit()

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 253:

Sabre\DAV\Server->invokeMethod()

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 321:

Sabre\DAV\Server->start()

/var/www/nextcloud/apps/dav/appinfo/v1/webdav.php - line 84:

Sabre\DAV\Server->exec()

/var/www/nextcloud/remote.php - line 167:

require_once("/var/www/nextcloud/apps/dav/appinfo/v1/webdav.php")`

**Comment by juliushaertl** [comment]
Pushed a fix to https://github.com/icewind1991/Streams/pull/15

**Comment by koelmel** [comment]
I believe, that the base problem is that different Icewind/Streams implementations are available under ./3rdparty/icewind and ./apps/files_external/3rdparty/icewind, as i see in my 20.0.9 installation. In my former nextcloud version 19.0.9 the files under the 2 paths were equal (for nextcloud 19.0.9 with files_external app version 1.10.0)
If the files_external app (in version 1.11.1 on 20.0.x) is activated, there are conflicts between both and the ClassLoader seems to take the last one, which requires because of the derivation from the WrapperHandler class that the Wrapper class has the wrapSource function without mode argument.
Nethertheless the adaptation of the WrapperHandler class should work, but @andyxheli has also found the problem that there are 2 implementations which are providing the WrapperHandler class (https://github.com/icewind1991/Streams/pull/15#issuecomment-847887505).

**Comment by andyxheli** [comment]
icewind1991/Streams#15 dose fix the issue on NC 21 but not on NC 20.0.10. For more info please see icewind1991/Streams#15

**Comment by koelmel** [comment]
the correct solution should be to clarify, which Icewind/Streams implementations either ./3rdparty/icewind/streams or ./apps/files_external/3rdparty/icewind/streams should be used. Different classes with the same namespaces should be avoided. Until the 20er version (i don't know which minor version has changed this) both Icewind/Streams paths are providing the same version.
I can confirm that deactivating the files_external app is a workaround for those who don't need this app.

**Comment by talesam** [comment]
I have version 21.0.2 stable and I have the same error.

```
{"reqId":"zq1vVhRg7dTooTiW19is","level":3,"time":"2021-06-27T22:53:57+00:00","remoteAddr":"198.98.60.69","user":"admin","app":"PHP","method":"PUT","url":"/remote.php/dav/files/admin/aaa/Readme.md","message":{"Exception":"Error","Message":"Undefined index: encryptedVersion at /var/www/html/lib/private/Files/Storage/Wrapper/Encryption.php#669","Code":0,"Trace":[{"file":"/var/www/html/lib/private/Files/Storage/Wrapper/Encryption.php","line":669,"function":"onError","class":"OC\\Log\\ErrorHandler","type":"::","args":[8,"Undefined index: encryptedVersion","/var/www/html/lib/private/Files/Storage/Wrapper/Encryption.php",669,{"0":"And 2 more entries, set log level to debug to see all entries","sourceStorage":"*** sensitive parameter replaced ***","sourceInternalPath":"*** sensitive parameter replaced ***","targetInternalPath":"*** sensitive parameter replaced ***","isRename":"*** sensitive parameter replaced ***","keepEncryptionVersion":"*** sensitive parameter replaced ***"}]},{"file":"/var/www/html/lib/private/Files/Storage/Wrapper/Encryption.php","line":767,"function":"updateEncryptedVersion","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/Files/Storage/Wrapper/Encryption.php","line":623,"function":"copyBetweenStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->","args":["*** sensitive parameter replaced ***","*** sensitive parameter replaced ***","*** sensitive parameter replaced ***","*** sensitive parameter replaced ***","*** sensitive parameter replaced ***"]},{"file":"/var/www/html/apps/files_trashbin/lib/Storage.php","line":240,"function":"moveFromStorage","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->","args":["*** sensitive parameter replaced ***","*** sensitive parameter replaced ***","*** sensitive parameter replaced ***"]},{"file":"/var/www/html/apps/dav/lib/Connector/Sabre/File.php","line":292,"function":"moveFromStorage","class":"OCA\\Files_Trashbin\\Storage","type":"->","args":["*** sensitive parameter replaced ***","*** sensitive parameter replaced ***","*** sensitive parameter replaced ***"]},{"file":"/var/www/html/apps/dav/lib/Connector/Sabre/Directory.php","line":156,"function":"put","class":"OCA\\DAV\\Connector\\Sabre\\File","type":"->","args":[null]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":1098,"function":"createFile","class":"OCA\\DAV\\Connector\\Sabre\\Directory","type":"->","args":["Readme.md",null]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/CorePlugin.php","line":504,"function":"createFile","class":"Sabre\\DAV\\Server","type":"->","args":["files/admin/aaa/Readme.md",null,null]},{"file":"/var/www/html/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"httpPut","class":"Sabre\\DAV\\CorePlugin","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":472,"function":"emit","class":"Sabre\\DAV\\Server","type":"->","args":["method:PUT",[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":253,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php","line":321,"function":"start","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/lib/Server.php","line":332,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args":[]},{"file":"/var/www/html/remote.php","line":167,"args":["/var/www/html/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/var/www/html/lib/private/Log/ErrorHandler.php","Line":92,"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36","version":"21.0.2.1","id":"60d9018cdb6fe"}
```

**Comment by andyxheli** [comment]
Issue fixed in 21.0.3
